<h2> All the answers for Part 1,2,3 is included here. </h2>

<h3> Part 1 </h3>

<h3> HTML Scraping </h3>

In [72]:
from bs4 import BeautifulSoup
import requests
import pandas as pd
import numpy as np

In [2]:
List_url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
source = requests.get(List_url).text

In [3]:
soup = BeautifulSoup(source, 'xml')

In [4]:
table=soup.find('table')

In [40]:
#dataframe will consist of three columns: PostalCode, Borough, and Neighborhood
column_names = ['Postalcode','Borough','Neighborhood']
df = pd.DataFrame(columns = column_names)

In [41]:
# Search all the postcode, borough, neighborhood 
for tr_cell in table.find_all('tr'):
    row_data=[]
    for td_cell in tr_cell.find_all('td'):
        row_data.append(td_cell.text.strip())
    if len(row_data)==3:
        df.loc[len(df)] = row_data

In [42]:
df.head()

Postalcode           Borough               Neighborhood
0        M1A      Not assigned                           
1        M2A      Not assigned                           
2        M3A        North York                  Parkwoods
3        M4A        North York           Victoria Village
4        M5A  Downtown Toronto  Regent Park, Harbourfront

<h2>Data Cleansing</h2>

Now we should drop the empty rows.
Looking at the data we can assume this.
If a cell in Neighborhood coloum doesn't have a value, then the cell in the Borought which has the same raw doesn't containg any value. So wed only have to drop raws considering a one coloum.
Let's drop raws considering Borought

In [43]:
df = df.dropna()
df = df[(df.Postalcode != 'Not assigned' ) & (df.Borough != 'Not assigned')]

In [44]:
df.head()

Postalcode           Borough                                 Neighborhood
2        M3A        North York                                    Parkwoods
3        M4A        North York                             Victoria Village
4        M5A  Downtown Toronto                    Regent Park, Harbourfront
5        M6A        North York             Lawrence Manor, Lawrence Heights
6        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

Let's check whether we have any repeating Postal values

In [45]:
df["Postalcode"].value_counts()

M1R    1
M4X    1
M8W    1
M1G    1
M2R    1
      ..
M9W    1
M5A    1
M1V    1
M1M    1
M3B    1
Name: Postalcode, Length: 103, dtype: int64

We're fine. Now make the index right.

In [46]:
df.reset_index(drop=True, inplace=True)

In [47]:
df.head()

Postalcode           Borough                                 Neighborhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [16]:
df.shape

(103, 3)

<h2>Part 2 </h2>

First, let's clone the above df dataframe

In [49]:
df2=df[:]

In [50]:
df2.head()

Postalcode           Borough                                 Neighborhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

The coordinates table

In [18]:
geo_df=pd.read_csv('http://cocl.us/Geospatial_data')

In [19]:
geo_df.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [20]:
geo_df.shape

(103, 3)

In [52]:
geo_df.rename(columns={'Postal Code':'Postalcode'},inplace=True)

In [53]:
geo_df.head()

Postalcode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

What we can do is merge those above table using the coloum 'Postalcode' and arrange them as given in question

In [54]:
geo_merged = pd.merge(geo_df, df2, on='Postalcode')

In [55]:
geo_merged.head()

Postalcode   Latitude  Longitude      Borough  \
0        M1B  43.806686 -79.194353  Scarborough   
1        M1C  43.784535 -79.160497  Scarborough   
2        M1E  43.763573 -79.188711  Scarborough   
3        M1G  43.770992 -79.216917  Scarborough   
4        M1H  43.773136 -79.239476  Scarborough   

                             Neighborhood  
0                          Malvern, Rouge  
1  Rouge Hill, Port Union, Highland Creek  
2       Guildwood, Morningside, West Hill  
3                                  Woburn  
4                               Cedarbrae

In [57]:
geo_data=geo_merged[['Postalcode','Borough','Neighborhood','Latitude','Longitude']]

In [58]:
geo_data.head()

Postalcode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Malvern, Rouge  43.806686   
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

<h2> Part 3 </h2>

In [97]:
import pandas as pd
import numpy as np
import requests
from bs4 import BeautifulSoup
import os
from sklearn.cluster import KMeans
import folium 
from geopy.geocoders import Nominatim 
import matplotlib.cm as cm
import matplotlib.colors as colors

In [59]:
CLIENT_ID = 'VVAVBHYK2VA1GRZ10J3SIDYYQOP2220W5NFYKDJPIKU4XQDI' # your Foursquare ID
CLIENT_SECRET = 'MF0UIKATTVOPSQLBLTAHYFFNBPFVJHSWFF2FEBACUIL13TKI' # your Foursquare Secret
VERSION = '20180604'

This is the plan. <br>
1)We can import JSON files according to the table we derived in Part 2 <br>
2)Then we create clusters by arranging them according to the venue categories <br>
3)We map them.

Let's define a function to get about 100 JSON files and apend the details as we get them. 

In [60]:
def getNearbyVenues(names, latitudes, longitudes):
    radius=500
    LIMIT=100
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Now, Let's call the above function to the each raw of the derived table from part 2

In [83]:
toronto_venues = getNearbyVenues(names=geo_data['Neighborhood'],
                                   latitudes=geo_data['Latitude'],
                                   longitudes=geo_data['Longitude']
                                  )

Malvern, Rouge
Rouge Hill, Port Union, Highland Creek
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
Kennedy Park, Ionview, East Birchmount Park
Golden Mile, Clairlea, Oakridge
Cliffside, Cliffcrest, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Wexford Heights, Scarborough Town Centre
Wexford, Maryvale
Agincourt
Clarks Corners, Tam O'Shanter, Sullivan
Milliken, Agincourt North, Steeles East, L'Amoreaux East
Steeles West, L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
York Mills, Silver Hills
Willowdale, Newtonbrook
Willowdale
York Mills West
Willowdale
Parkwoods
Don Mills
Don Mills
Bathurst Manor, Wilson Heights, Downsview North
Northwood Park, York University
Downsview
Downsview
Downsview
Downsview
Victoria Village
Parkview Hill, Woodbine Gardens
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto
The Danforth West, Riverdale
India Bazaar, The Beaches West
Studio District
Lawrence P

In [84]:
toronto_venues.head()

Neighborhood  Neighborhood Latitude  \
0                          Malvern, Rouge              43.806686   
1  Rouge Hill, Port Union, Highland Creek              43.784535   
2  Rouge Hill, Port Union, Highland Creek              43.784535   
3       Guildwood, Morningside, West Hill              43.763573   
4       Guildwood, Morningside, West Hill              43.763573   

   Neighborhood Longitude                                              Venue  \
0              -79.194353                                            Wendy’s   
1              -79.160497                              Royal Canadian Legion   
2              -79.160497  SEBS Engineering Inc. (Sustainable Energy and ...   
3              -79.188711                                     RBC Royal Bank   
4              -79.188711                                  G & G Electronics   

   Venue Latitude  Venue Longitude              Venue Category  
0       43.807448       -79.199056        Fast Food Restaurant  
1       43.782533       -79.163085                         Bar  
2       43.782371       -79.156820  Construction & Landscaping  
3       43.766790       -79.191151                        Bank  
4       43.765309       -79.191537           Electronics Store

In [85]:
toronto_venues.shape

(2137, 7)

Let's first check what are the categories of Neighborhood we have and how many in each

In [86]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                             
Agincourt                                                            4   
Alderwood, Long Branch                                              10   
Bathurst Manor, Wilson Heights, Downsview North                     19   
Bayview Village                                                      4   
Bedford Park, Lawrence Manor East                                   22   
...                                                                ...   
Willowdale, Newtonbrook                                              1   
Woburn                                                               3   
Woodbine Heights                                                     9   
York Mills West                                                      4   
York Mills, Silver Hills                                             1   

                                                 Neighborhood Longitude  \
Neighborhood                                                              
Agincourt                                                             4   
Alderwood, Long Branch                                               10   
Bathurst Manor, Wilson Heights, Downsview North                      19   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    22   
...                                                                 ...   
Willowdale, Newtonbrook                                               1   
Woburn                                                                3   
Woodbine Heights                                                      9   
York Mills West                                                       4   
York Mills, Silver Hills                                              1   

                                                 Venue  Venue Latitude  \
Neighborhood                                                             
Agincourt                                            4               4   
Alderwood, Long Branch                              10              10   
Bathurst Manor, Wilson Heights, Downsview North     19              19   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   22              22   
...                                                ...             ...   
Willowdale, Newtonbrook                              1               1   
Woburn                                               3               3   
Woodbine Heights                                     9               9   
York Mills West                                      4               4   
York Mills, Silver Hills                             1               1   

                                                 Venue Longitude  \
Neighborhood                                                       
Agincourt                                                      4   
Alderwood, Long Branch                                        10   
Bathurst Manor, Wilson Heights, Downsview North               19   
Bayview Village                                                4   
Bedford Park, Lawrence Manor East                             22   
...                                                          ...   
Willowdale, Newtonbrook                                        1   
Woburn                                                         3   
Woodbine Heights                                               9   
York Mills West                                                4   
York Mills, Silver Hills                                       1   

                                                 Venue Category  
Neighborhood                                                     
Agincourt                                                     4  
Alderwood, Long Branch                                       10 

Okay. Now, we have to find what types of Venues are in the Neighborhood. SO first, let's one hot encode the full table using venue Category and group them accordinf to the Neighborhood

In [87]:
# one hot encoding. Apllied to the full table with 1237 raws
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
toronto_onehot.drop(['Neighborhood'],axis=1,inplace=True) 
toronto_onehot.insert(loc=0, column='Neighborhood', value=toronto_venues['Neighborhood'] )
toronto_onehot.head()

Neighborhood  Accessories Store  \
0                          Malvern, Rouge                  0   
1  Rouge Hill, Port Union, Highland Creek                  0   
2  Rouge Hill, Port Union, Highland Creek                  0   
3       Guildwood, Morningside, West Hill                  0   
4       Guildwood, Morningside, West Hill                  0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   ...  Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0  ...              0                              0                 0   
1  ...              0                              0                 0   
2  ...              0                              0                 0   
3  ...              0                              0                 0   
4  ...              0                              0                 0   

   Video Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0            0                      0                0         0            0   
1            0                      0                0         0            0   
2            0                      0                0         0            0   
3            0                      0                0         0            0   
4            0                      0                0         0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 266 columns]

In [88]:
toronto_onehot.shape

(2137, 266)

Now we can group them

In [89]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                0.0      0.0                 0.0           0.0   
1                0.0      0.0                 0.0           0.0   
2                0.0      0.0                 0.0           0.0   
3                0.0      0.0                 0.0           0.0   
4                0.0      0.0                 0.0           0.0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0             0.0              0.0               0.0             0.000000   
1             0.0              0.0               0.0             0.000000   
2             0.0              0.0               0.0             0.000000   
3             0.0              0.0               0.0             0.000000   
4             0.0              0.0               0.0             0.045455   

   ...  Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0  ...            0.0                            0.0               0.0   
1  ...            0.0                            0.0               0.0   
2  ...            0.0                            0.0               0.0   
3  ...            0.0                            0.0               0.0   
4  ...            0.0                            0.0               0.0   

   Video Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0     0.000000                    0.0              0.0       0.0          0.0   
1     0.000000                    0.0              0.0       0.0          0.0   
2     0.052632                    0.0              0.0       0.0          0.0   
3     0.000000                    0.0              0.0       0.0          0.0   
4     0.000000                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 266 columns]

In [99]:
toronto_grouped.shape

(96, 266)

We have 266 Venue categories here. Let's check what are the 10 most common

In [100]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [101]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    columns.append('Number {} Most-Common Venue'.format(ind+1))
    #try:
    #    columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    #except:
    #    columns.append('{}th Most Common Venue'.format(ind+1))
# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood Number 1 Most-Common Venue  \
0                                        Agincourt  Latin American Restaurant   
1                           Alderwood, Long Branch                Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                Coffee Shop   
3                                  Bayview Village        Japanese Restaurant   
4                Bedford Park, Lawrence Manor East         Italian Restaurant   

  Number 2 Most-Common Venue Number 3 Most-Common Venue  \
0               Skating Rink                     Lounge   
1                        Gym                Coffee Shop   
2                       Bank  Middle Eastern Restaurant   
3                       Café                       Bank   
4             Sandwich Place                 Restaurant   

  Number 4 Most-Common Venue Number 5 Most-Common Venue  \
0             Breakfast Spot                    Dog Run   
1               Dance Studio               Skating Rink   
2                Pizza Place             Sandwich Place   
3         Chinese Restaurant           Department Store   
4                Coffee Shop                    Butcher   

  Number 6 Most-Common Venue Number 7 Most-Common Venue  \
0               Dessert Shop         Dim Sum Restaurant   
1                        Pub                   Pharmacy   
2                      Diner                Bridal Shop   
3         Dim Sum Restaurant                      Diner   
4                       Café          Indian Restaurant   

  Number 8 Most-Common Venue Number 9 Most-Common Venue  \
0                      Diner             Discount Store   
1                       Pool             Sandwich Place   
2                 Restaurant              Deli / Bodega   
3             Discount Store        Distribution Center   
4           Sushi Restaurant                Pizza Place   

  Number 10 Most-Common Venue  
0         Distribution Center  
1                Dessert Shop  
2              Ice Cream Shop  
3                 Yoga Studio  
4                    Pharmacy

In [92]:
# set number of clusters to 5
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [93]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = geo_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head()

Postalcode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Malvern, Rouge  43.806686   
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster Labels       1st Most Common Venue  \
0 -79.194353             1.0        Fast Food Restaurant   
1 -79.160497             0.0  Construction & Landscaping   
2 -79.188711             0.0         Rental Car Location   
3 -79.216917             0.0                 Coffee Shop   
4 -79.239476             0.0                 Gas Station   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0           Yoga Studio          Dance Studio     Electronics Store   
1                   Bar           Yoga Studio               Dog Run   
2    Mexican Restaurant        Breakfast Spot        Medical Center   
3     Korean Restaurant           Yoga Studio               Dog Run   
4       Thai Restaurant   Fried Chicken Joint                  Bank   

         5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0  Eastern European Restaurant             Drugstore            Donut Shop   
1                 Dessert Shop    Dim Sum Restaurant                 Diner   
2                 Intersection                  Bank     Electronics Store   
3                 Dessert Shop    Dim Sum Restaurant                 Diner   
4           Athletics & Sports  Caribbean Restaurant                Bakery   

  8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
0      Doner Restaurant               Dog Run          Distribution Center  
1        Discount Store   Distribution Center             Doner Restaurant  
2     Convenience Store     Convention Center  Eastern European Restaurant  
3        Discount Store   Distribution Center                   Donut Shop  
4      Hakka Restaurant      Doner Restaurant                      Dog Run

In [102]:
neighborhoods_venues_sorted.head()

Neighborhood Number 1 Most-Common Venue  \
0                                        Agincourt  Latin American Restaurant   
1                           Alderwood, Long Branch                Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                Coffee Shop   
3                                  Bayview Village        Japanese Restaurant   
4                Bedford Park, Lawrence Manor East         Italian Restaurant   

  Number 2 Most-Common Venue Number 3 Most-Common Venue  \
0               Skating Rink                     Lounge   
1                        Gym                Coffee Shop   
2                       Bank  Middle Eastern Restaurant   
3                       Café                       Bank   
4             Sandwich Place                 Restaurant   

  Number 4 Most-Common Venue Number 5 Most-Common Venue  \
0             Breakfast Spot                    Dog Run   
1               Dance Studio               Skating Rink   
2                Pizza Place             Sandwich Place   
3         Chinese Restaurant           Department Store   
4                Coffee Shop                    Butcher   

  Number 6 Most-Common Venue Number 7 Most-Common Venue  \
0               Dessert Shop         Dim Sum Restaurant   
1                        Pub                   Pharmacy   
2                      Diner                Bridal Shop   
3         Dim Sum Restaurant                      Diner   
4                       Café          Indian Restaurant   

  Number 8 Most-Common Venue Number 9 Most-Common Venue  \
0                      Diner             Discount Store   
1                       Pool             Sandwich Place   
2                 Restaurant              Deli / Bodega   
3             Discount Store        Distribution Center   
4           Sushi Restaurant                Pizza Place   

  Number 10 Most-Common Venue  
0         Distribution Center  
1                Dessert Shop  
2              Ice Cream Shop  
3                 Yoga Studio  
4                    Pharmacy

In [116]:
address = 'Toronto, CA'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Manhattan are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Manhattan are 43.6534817, -79.3839347.


In [120]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    print('Cluster is ', cluster)
    print('lat is ', lat)
    print('poi is ', poi)
    
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        #color=rainbow[int(cluster+2)],
        fill=True,
        fill_color=rainbow[cluster-1]
        #fill_color=rainbow[int(cluster+2)],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters